src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp

```diff
@@ -125,107 +131,129 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(std::shared_ptr<CryptoParamete
                             if (qPrev == moduliQ[j]) {
                                 hasSameMod = true;
+                                break;
                             }
                         }
                     }
                     moduliQ[i] = qPrev;
                 }
                 else {
-                    NativeInteger sfInt = std::llround(sf);
-                    NativeInteger sfRem = sfInt.Mod(cyclOrder);
                     NativeInteger qNext = sfInt + NativeInteger(cyclOrder) - sfRem + NativeInteger(1);
-                    bool hasSameMod     = true;
                     while (hasSameMod) {
                         hasSameMod = false;
                         qNext      = NextPrime(qNext, cyclOrder);
-                        for (uint32_t j = i + 1; j < numPrimes; j++) {
+                        for (size_t j = i + 1; j < numPrimes; j++) {
                             if (qNext == moduliQ[j]) {
                                 hasSameMod = true;
+                                break;
                             }
                         }
                     }
                     moduliQ[i] = qNext;
                 }
+                if (moduliQ[i] > maxPrime)
+                    maxPrime = moduliQ[i];
+                else if (moduliQ[i] < minPrime)
+                    minPrime = moduliQ[i];
 
                 rootsQ[i] = RootOfUnity(cyclOrder, moduliQ[i]);
-                cnt++;
             }
         }
     }
 
     if (firstModSize == dcrtBits) {  // this requires dcrtBits < 60
-        moduliQ[0] = PreviousPrime<NativeInteger>(qPrev, cyclOrder);
+        moduliQ[0] = NextPrime<NativeInteger>(maxPrime, cyclOrder);
     }
     else {
         moduliQ[0] = LastPrime<NativeInteger>(firstModSize, cyclOrder);
+
+        // find if the value of moduliQ[0] is already in the vector starting with moduliQ[1] and
+        // if there is, then get another prime for moduliQ[0]
+        const auto pos = std::find(moduliQ.begin() + 1, moduliQ.end(), moduliQ[0]);
+        if (pos != moduliQ.end()) {
+            moduliQ[0] = NextPrime<NativeInteger>(maxPrime, cyclOrder);
+        }
     }
+    if (moduliQ[0] > maxPrime)
+        maxPrime = moduliQ[0];
+
     rootsQ[0] = RootOfUnity(cyclOrder, moduliQ[0]);
 
     if (scalTech == FLEXIBLEAUTOEXT) {
-        // no need for extra checking as extraModSize is automatically chosen by the library
-        moduliQ[numPrimes] = FirstPrime<NativeInteger>(extraModSize - 1, cyclOrder);
-        rootsQ[numPrimes]  = RootOfUnity(cyclOrder, moduliQ[numPrimes]);
+        // find if the value of moduliQ[numPrimes] is already in the vector.
+        // in this case we must redefine the position of "end()" as "end()-1"
+        // currently "moduliQ[numPrimes] == 0" as it is not set up anywhere in this code yet
+        const auto endPos = moduliQ.end() - 1;
+        auto pos          = std::find(moduliQ.begin(), endPos, moduliQ[numPrimes]);
```

Comment:
I don't understand what this line does. Why are we looking for a value = 0 = moduliQ[numPrimes] in the vector of modulus between 0 and numPrimes - 1?

Reply:
will change and add extra comments